Blur media from unfollowed

kernelkind opened the pull request:
closes #679
closes #797

jb55 commented:
looks like it needs rebase ?

kernelkind commented:
rebased with master

kernelkind commented:
rebased with master

kernelkind commented:
I think it makes more sense to put the promise inside the `Job`

jb55 commented on crates/notedeck_columns/src/contacts.rs:
```diff
@@ -0,0 +1,62 @@
+    Filter::new().authors([pk]).kinds([3]).limit(1).build()
+}
+
+fn note_follows(contacts_note: Note<'_>, pk: &[u8; 32]) -> bool {
```
it's funny, normally I would have said this is too slow but these are all non-copy iterators over compact binary data. so it's likely fast enough until we have the follow cache.

jb55 commented on crates/notedeck_columns/src/contacts.rs:
```diff
@@ -0,0 +1,62 @@
+            continue;
+        }
+
+        let Some(author) = tag.get_unchecked(1).variant().id() else {
```
can just use tag.get_id() here

jb55 commented on crates/notedeck_columns/src/contacts.rs:
```diff
@@ -0,0 +1,62 @@
+            continue;
+        }
+
+        let Some(t) = tag.get_unchecked(0).variant().str() else {
```
can just use tag.get_str() here

jb55 reviewed (approved):
looks good!

jb55 commented:
needs rebase

kernelkind commented:
> I'm not entirely sure what the point of the jobs thing is, this seems like its more of an image cache thing? I don't think we should be exposing it to all of our notedeck app interfaces either (AppContext, etc)

Take another look at it, now it accepts an async closure and does not copy unless needed to call the closure. Also, it limits the amount of concurrently running threads, defaulted to 10.

And yeah on second thought it would probably make more sense moving it into `Damus` instead of `Notedeck`.

Let me know what you think

kernelkind commented:
rebased with master

jb55 commented:
unfortunately trust_media_from_pk2 is just way too slow without the author_kind query plan implementation , as I suspected:

![Mar23-115717](https://github.com/user-attachments/assets/dc1fa797-4cda-4a9d-b929-809afe5ad5ef)

![Mar23-120029](https://github.com/user-attachments/assets/f4def6c8-8d28-45a3-adf1-5e58051d8605)

I will quickly implement that in nostrdb and see if it fixes this

jb55 commented:
its faster but still pretty slow. We'll need to update it to only call trust_media_from_pk2 if there is media.

kernelkind commented:
moved `Jobs` into `Damus`

kernelkind commented:
rebased with master

kernelkind commented:
> its faster but still pretty slow. We'll need to update it to only call trust_media_from_pk2 if there is media.

how do i use the faster way in nostrdb to get a user's contacts note?

kernelkind commented:
addressed:
> its faster but still pretty slow. We'll need to update it to only call trust_media_from_pk2 if there is media.

and rebased with master

jb55 commented:
On Sun, Apr 13, 2025 at 12:50:56PM -0700, kernelkind wrote:
>kernelkind left a comment (damus-io/notedeck#779)
>
>> its faster but still pretty slow. We'll need to update it to only call trust_media_from_pk2 if there is media.
>
>how do i use the faster way in nostrdb to get a user's contacts note?

its as fast as possible at the moment, when I mean faster we will likely
need a persistent who-follows-who cache so we don't have to scan contact
lists all the time. I can add this to nostrdb so that its automatically
calculated when processing contact lists.

kernelkind commented:
rebased with master

kernelkind commented:
Took a bit longer than I thought. But now obfuscation will appear for when the actual image is still loading in general, in addition to having obfuscation default for unfollowed and having shimmers on loading images. Here is a demonstration:

https://v.nostr.build/GVaE8jmOSlbEibK1.mp4

kernelkind commented:
rebased with master
